File: scripts/cloud/latest_run.py
"""Resolve the latest Extra High run for wait-notify / FLEET_DONE.

Leftover FINISHED is not terminal while a newer run is CREATING or RUNNING.
GET /v1/agents/{id}/runs (collection) is the source of latest, not a pinned
leftover run id or a stale agent latestRunId.
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

IN_FLIGHT = frozenset({"CREATING", "RUNNING"})
TERMINAL = frozenset({"FINISHED", "ERROR", "CANCELLED", "EXPIRED"})


def run_status(run: dict[str, Any] | None) -> str:
    if not isinstance(run, dict):
        return ""
    return str(run.get("status") or run.get("runStatus") or "").strip().upper()


def unwrap_entity(data: Any, key: str) -> Any:
    if isinstance(data, dict) and key in data and "id" not in data:
        inner = data[key]
        if isinstance(inner, dict):
            return inner
    return data


def unwrap_runs(payload: Any) -> list[dict[str, Any]]:
    """Extract run dicts from GET /v1/agents/{id}/runs."""
    if isinstance(payload, list):
        items: Any = payload
    elif isinstance(payload, dict):
        items = payload.get("items") or payload.get("runs") or []
        if not items and payload.get("id") and (
            "status" in payload or "runStatus" in payload
        ):
            items = [payload]
    else:
        items = []
    if not isinstance(items, list):
        return []
    out: list[dict[str, Any]] = []
    for item in items:
        if not isinstance(item, dict):
            continue
        run = unwrap_entity(item, "run")
        if isinstance(run, dict) and (run.get("id") or run.get("status") or run.get("runStatus")):
            out.append(run)
    return out


def created_at_ms(run: dict[str, Any]) -> int:
    for key in ("createdAtMs", "created_at_ms"):
        val = run.get(key)
        if isinstance(val, (int, float)) and val > 0:
            return int(val)
    for key in ("createdAt", "created_at"):
        val = run.get(key)
        if isinstance(val, (int, float)):
            num = int(val)
            return num if num > 1_000_000_000_000 else num * 1000
        if isinstance(val, str) and val.strip():
            raw = val.strip().replace("Z", "+00:00")
            try:
                dt = datetime.fromisoformat(raw)
            except ValueError:
                continue
            if dt.tzinfo is None:
                dt = dt.replace(tzinfo=timezone.utc)
            return int(dt.timestamp() * 1000)
    return 0


def pick_latest_run(runs: list[dict[str, Any]]) -> dict[str, Any] | None:
    if not runs:
        return None
    indexed = list(enumerate(runs))
    indexed.sort(key=lambda pair: (created_at_ms(pair[1]), pair[0]))
    return indexed[-1][1]


def waiter_observe(
    runs: list[dict[str, Any]],
    pinned: dict[str, Any] | None = None,
) -> dict[str, Any] | None:
    """Run wait-notify should treat as current.

    In-flight CREATING/RUNNING always wins over leftover FINISHED, including
    when --run / latestRunId still names the leftover.
    """
    combined: list[dict[str, Any]] = list(runs)
    if isinstance(pinned, dict):
        pid = str(pinned.get("id") or "")
        if pid and not any(str(row.get("id") or "") == pid for row in combined):
            combined.append(pinned)
    in_flight = [row for row in combined if run_status(row) in IN_FLIGHT]
    if in_flight:
        return pick_latest_run(in_flight)
    return pick_latest_run(combined)


def may_fleet_done(run: dict[str, Any] | None) -> bool:
    return bool(run) and run_status(run) in TERMINAL

File: scripts/cloud/sdk/latest_run.ts
/** Latest Extra High run for wait-notify / FLEET_DONE.
 *
 * Leftover FINISHED is not terminal while a newer run is CREATING or RUNNING.
 * Distinct from occupancy #132 and paginated-catalog. Never Bot CloudAgent.
 */

export const IN_FLIGHT = new Set(["CREATING", "RUNNING"]);
export const TERMINAL = new Set(["FINISHED", "ERROR", "CANCELLED", "EXPIRED"]);

export type RunLike = {
  id?: string;
  status?: string;
  runStatus?: string;
  createdAt?: number | string;
  createdAtMs?: number;
  created_at?: number | string;
  created_at_ms?: number;
};

export function runStatus(run: RunLike | undefined): string {
  if (!run) return "";
  const raw = run.status || run.runStatus || "";
  return String(raw).trim().toUpperCase();
}

export function unwrapRuns(payload: unknown): RunLike[] {
  if (Array.isArray(payload)) {
    return payload.filter(isRunLike);
  }
  if (!payload || typeof payload !== "object") return [];
  const rec = payload as Record<string, unknown>;
  const nested = rec.items ?? rec.runs;
  if (Array.isArray(nested)) {
    return nested.flatMap((item) => {
      if (!item || typeof item !== "object") return [];
      const row = item as Record<string, unknown>;
      if (row.run && typeof row.run === "object" && !Array.isArray(row.run) && !("id" in row)) {
        return isRunLike(row.run) ? [row.run as RunLike] : [];
      }
      return isRunLike(item) ? [item as RunLike] : [];
    });
  }
  if (isRunLike(payload)) return [payload];
  return [];
}

function isRunLike(value: unknown): value is RunLike {
  if (!value || typeof value !== "object") return false;
  const row = value as RunLike;
  return Boolean(row.id || row.status || row.runStatus);
}

export function createdAtMs(run: RunLike): number {
  const msKeys = [run.createdAtMs, run.created_at_ms];
  for (const val of msKeys) {
    if (typeof val === "number" && val > 0) return Math.trunc(val);
  }
  const raw = run.createdAt ?? run.created_at;
  if (typeof raw === "number" && Number.isFinite(raw)) {
    const num = Math.trunc(raw);
    return num > 1_000_000_000_000 ? num : num * 1000;
  }
  if (typeof raw === "string" && raw.trim()) {
    const ms = Date.parse(raw.trim());
    if (!Number.isNaN(ms)) return ms;
  }
  return 0;
}

export function pickLatestRun(runs: RunLike[]): RunLike | undefined {
  if (!runs.length) return undefined;
  let best = 0;
  for (let i = 1; i < runs.length; i += 1) {
    const a = createdAtMs(runs[best]);
    const b = createdAtMs(runs[i]);
    if (b > a || (b === a && i > best)) best = i;
  }
  return runs[best];
}

export function waiterObserve(runs: RunLike[], pinned?: RunLike): RunLike | undefined {
  const combined = [...runs];
  if (pinned) {
    const pid = String(pinned.id || "");
    if (pid && !combined.some((row) => String(row.id || "") === pid)) {
      combined.push(pinned);
    }
  }
  const inFlight = combined.filter((row) => IN_FLIGHT.has(runStatus(row)));
  if (inFlight.length) return pickLatestRun(inFlight);
  return pickLatestRun(combined);
}

export function mayFleetDone(run: RunLike | undefined): boolean {
  return Boolean(run) && TERMINAL.has(runStatus(run));
}
